### PR TITLE
Enumerate Dates instead of Times when DTSTART is a Date. Closes #13

### LINKED ISSUE #13: Feature request: RRule::Rule.new to accept Date (without time) for dtstart
A caldav server can return an event whose DTSTART and DTEND do not indicate a time or offset. This is commonly the case in icloud's server, where the event is a holiday:
```
DTEND;VALUE=DATE:20200120
DTSTART;VALUE=DATE:20190120
RRULE:FREQ=YEARLY
SUMMARY:My birthday
```
`RRule::Rule`'s constructor expects `dtstart` to have `#to_i`, which a `Date` does not have, and anyway calls `#in_time_zone`, which converts it to a `DateTime` at midnight in the current timezone. It would be nice to have the ability to use `RRule::Rule#between` to expand such an event, with the output events also having dates without times, so that each occurrence produced can also be a date-without-time.

### PR DESCRIPTION
**Pull request by leoarnold**

### CODE REVIEW

File: lib/rrule/rule.rb [hunk 1]
```diff
@@ -111,7 +111,9 @@ def parse_options(rule)
           raise InvalidRRule, 'COUNT must be a non-negative integer' if i < 0
           options[:count] = i
         when 'UNTIL'
-          options[:until] = Time.parse(value)
+          # The value of the UNTIL rule part MUST have the same
+          # value type as the "DTSTART" property.
+          options[:until] = @dtstart.is_a?(Date) ? Date.parse(value) : Time.parse(value)
```
Comment (nerdrew) [inline lib/rrule/rule.rb]: It is a little awkward that `Date.parse` won't throw an error if `value` is a time-like string...

```ruby
Date.parse('2020-01-01T12:34:01Z') # => #<Date: 2020-01-01 ((2458850j,0s,0n),+0s,2299161j)>
```

Though, I'm not sure this matters too much.

**Comment by ordermuppet** [comment]
Thanks for these fixes!

File: lib/rrule/rule.rb [hunk 2]
```diff
@@ -154,7 +156,11 @@ def parse_options(rule)
 
       options[:byweekday], options[:bynweekday] = options[:byweekday].partition { |wday| wday.ordinal.nil? } unless options[:byweekday].nil?
 
-      options[:timeset] = [{ hour: (options[:byhour].presence || dtstart.hour), minute: (options[:byminute].presence || dtstart.min), second: (options[:bysecond].presence || dtstart.sec) }]
+      # The BYSECOND, BYMINUTE and BYHOUR rule parts MUST NOT be specified
+      # when the associated "DTSTART" property has a DATE value type.
+      # These rule parts MUST be ignored in RECUR value that violate the
+      # above requirement
```
Comment (nerdrew) [inline lib/rrule/rule.rb]: Is this comment from the spec?

File: lib/rrule/generators/generator.rb [hunk 3]
```diff
@@ -9,6 +9,8 @@ def initialize(context)
     end
 
     def process_timeset(date, timeset)
+      return date if timeset.blank?
```
Comment (nerdrew) [inline lib/rrule/generators/generator.rb]: It looks like this is always called from within a `.flat_map`, but it feels weird that the method can now return a `Date` or an `Array`.

File: lib/rrule/rule.rb (hunk 2 above)
Comment (leoarnold) [reply lib/rrule/rule.rb]: Yes, verbatim.

File: lib/rrule/generators/generator.rb (hunk 3 above)
Comment (leoarnold) [reply lib/rrule/generators/generator.rb]: Good point. I did not notice. I'll take a look at that next time around.

**Pull request merged**